utilities/include/vec3.hpp

```diff
@@ -1,212 +1,60 @@
 #ifndef VEC3_HPP
 #define VEC3_HPP
 
-
 #include <iostream>
-#include <cmath> //constexpr is compiler specific
-class vec3{
-    float x,y,z;
-    public:
-        constexpr vec3() : x {0}, y {0}, z {0} {}
-
-        //Cartesian coordinate Constructor
-        constexpr vec3(float X,float Y,float Z) : x {X}, y {Y}, z {Z} {}
-
-        //norm function
-        #ifndef __GNUC__
-
-        float norm() const {
-            return sqrt(x * x + y * y + z * z);
-        }
-
-        #endif
-
-        #ifdef __GNUC__
-        
-        constexpr float norm() const {
-            return sqrt(x * x + y * y + z * z);
-        }
+#include <cmath>
 
-        #endif
-
-
-
-        //unit vector function
-        constexpr vec3 unit() const{
-            return *this/this->norm();
-        }
-
-        //getters
-        constexpr float get_X() const{
-            return x;
-        }
-
-        constexpr float get_Y() const{
-            return y;
-        }
-        
-        constexpr float get_Z() const{
-            return z;
-        }
+#include "vec4.hpp"
+#include "config.hpp"
 
-        //vector addition
-        friend constexpr vec3 operator+(const vec3& v1, const vec3& v2);
+template <class T>
+struct as_index
```

Comment:
Also it would be better if we can encapsulate `as_index` and `indices` inside `Vec3`, I tried but ran into errors

Reply:
Yes, I prefer the nested class idea. We can have these structs as private nested classes so that outsiders cannot use them (accidentally or otherwise). Also, the `cross` function cannot be constexpr in its current form.

Reply:
Do you mean we have to use `std::is_constant_evaluated` here?

Reply:
You should add a `std::is_constant_evaluated` branch on every single math function, yes.